TST: Improve tests for numpy.pad

lagru opened the pull request:
This aims to improve the test suite for numpy.pad in anticipation for #11358. E.g. refactor the classes ValueError1, ValueError2, ..., add some missing edge cases and  group tests that cover the same aspect of pad. e67ecc3 actually removes ineffective unit tests.

lagru committed:
TST: Move test for negative stat_length

and extend coverage to all modes and more variations of a negative
stat_length.

lagru committed:
TST: Merge tests for pad_width in single class

lagru committed:
TST: Test behavior of pad's kwargs for all modes

lagru committed:
TST: Move test to TestReflect

Can be grouped with already existing test class checking the behavior
for the reflect mode.

lagru committed:
TST: Simplify regression test for object input

lagru committed:
TST: Move testing pad_width as ndarray

Can be grouped in class TestPadWidth as this test checks if an ndarray
is accepted as the value to pad_width

lagru committed:
TST: Remove faulty tests for pad_width's type

These test were ineffective. The TypeError raised in these test was not
actually due to pad_width receiving the wrong type but due to the
missing parameter mode.

Added missing type complex to the appropriate existing test checking for
pad_widths type behavior.

lagru commented on numpy/lib/tests/test_arraypad.py:
```diff
@@ -12,6 +12,20 @@
 from numpy.lib.arraypad import _as_pairs
 
 
+_all_modes = {
```
At some point it made sense to introduce this global variable to avoid duplication.

lagru commented on numpy/lib/tests/test_arraypad.py:
```diff
@@ -1122,8 +1155,75 @@ def _padwithtens(vector, pad_width, iaxis, kwargs):
+        ((1, 2), (3, 4), (5, 6)),
+        ((3, 4, 5), (0, 1, 2)),
+    ])
+    @pytest.mark.parametrize("mode", _all_modes.keys())
```
Using `parametrize` instead of for loops makes the code more readable but really bloats the number of unit tests and thus the output of pytest. What's the general preference here?

seberg replied:
I think we tend to use parameterize now. If it really does bloat output too much, maybe we can also ask pytest to add some nobs to help.

lagru replied:
Okay, good to know.

lagru commented:
Oh, and reviewing is probably easiest if you follow the commits one by one. Most are independent of another.

seberg commented on numpy/lib/tests/test_arraypad.py:
```diff
@@ -1176,20 +1190,49 @@ def test_object_input(self):
+    @pytest.mark.parametrize(
+        "pad_width", ["3", "word", None, object(), 3.4, ((2, 3, 4), (3, 2))])
+    @pytest.mark.parametrize("mode", _all_modes.keys())
+    def test_bad_pad_with_type(self, pad_width, mode):
```
typo -> width, no worries though

seberg reviewed (approved):
Looks very thorough to me. Do you want to add morethings, or should we merge this?

lagru commented:
@seberg Thanks for looking at this. It's been some time but I remember that there were a few thinks I wanted to look at. I'll do so in the next few days and try to finish this.

lagru committed:
TST: Move test for pad_width of zero

lagru committed:
TST: Move test for simple stat_length

lagru committed:
TST: Simplify classes with only one test

lagru committed:
TST: Add naive test for non-contiguous arrays

lagru committed:
MAINT: Don't import pad directly

Using np.pad instead of directly importing the function seems to be more
inline with other test modules.

lagru committed:
STY: Make class layout consistent in module

lagru commented:
@seberg I think this is ready to go. Test coverage is nearly at 100%. Not covered are:

* https://github.com/numpy/numpy/blob/dea85807c258ded3f75528cce2a444468de93bc1/numpy/lib/arraypad.py#L961 Not really sure what this function does and how to fix that.

* https://github.com/numpy/numpy/blob/dea85807c258ded3f75528cce2a444468de93bc1/numpy/lib/arraypad.py#L1288-L1292
I think this is a really wird bug. Coverage.py complains that the if statement is never false and therefore never jumped to the `continue` statement which is plain wrong. If I add a statement (e.g. `_ = 1`) in front of the continue statement the bug disappears...

* https://github.com/numpy/numpy/blob/dea85807c258ded3f75528cce2a444468de93bc1/numpy/lib/arraypad.py#L1337 Is never false. Can be ignored as it will be obsolete with the rewrite anyway.

Also I couldn't figure out how to use `runtests.py` for this purpose. I haven't managed to use the `-t` option to restrict the selected tests (the script complains with `ERROR: file or package not found`) nor how to use the `--coverage` option (prints `runtests.py: error: unrecognized arguments: --cov-report=html:/**/numpy/build/coverage --cov=/**/numpy/numpy`)

mattip commented:
`-t` specifies the path to the python file, i.e. `-t numpy/core/tests/test_multiarray.py`

`--cov` needs [`pytest-cov`](https://pypi.org/project/pytest-cov/)

mattip commented on numpy/lib/tests/test_arraypad.py:
```diff
@@ -1142,121 +1225,39 @@ def test_check_simple(self):
+def test_missing_mode():
+    match = r"pad\(\) missing 1 required positional argument: 'mode'"
+    with pytest.raises(TypeError, match=match):
+        np.pad(np.ones((5, 6)), 4)
```
tests are failing, it seems `mode` is required here?

lagru replied:
That's actually intended with this test. The problem seems to be that `match` doesn't match because the returned message leads with `_pad_dispatcher() missing 1 required positional [...]` instead of `pad()`. Should I simply change the match-string or is this actually a bug (e.g. `_pad_dispatcher` not correctly wrapping? When I try this in my console the returned error leads with `pad()`.

mattip replied:
Just start the match with "missing 1 ...". The `_pad_dispatcher` is another problem, but out of scope for this PR, pinging @shoyer to take a look

shoyer replied:
See https://github.com/numpy/numpy/issues/12028#issuecomment-464176722

lagru commented:
@mattip Thanks! It seems like for some reason I need to specify an absolute path for the `-t` option; relative won't work for some reason.

lagru committed:
TST: Fix match-string for missing pad mode error

The CLI fails due to error message containing a reference to
_pad_dispatcher() being returned instead of pad(). For some reason this
test passes when run locally.

seberg commented:
Thanks @lagru! Nice work.

lagru commented:
You're welcome. :slightly_smiling_face:  I'll merge these changes into #11358 tomorrow and have a look at what is left to do.
